Implement parsing of CSS font-variant-alternates

mdubet opened the pull request:
#### 96c123ec8040c9890edb7c5cc8846b8026108ea0
<pre>
Implement parsing of CSS font-variant-alternates
<a href="https://bugs.webkit.org/show_bug.cgi?id=149772">https://bugs.webkit.org/show_bug.cgi?id=149772</a>
rdar://problem/22959741

Reviewed by Myles C. Maxfield.

This patch implements the complete parsing of the property.
Only the &quot;normal&quot; and &quot;historical-forms&quot; values are actually
passed down to the platform.

<a href="https://www.w3.org/TR/css-fonts-4/#font-variant-alternates-prop">https://www.w3.org/TR/css-fonts-4/#font-variant-alternates-prop</a>

* LayoutTests/fast/text/font-variant-shorthand-expected.txt:
* LayoutTests/fast/text/font-variant-shorthand.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-variant-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-001-expected.txt:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSFontVariantAlternatesValue.cpp: Added.
(WebCore::CSSFontVariantAlternatesValue::CSSFontVariantAlternatesValue):
(WebCore::CSSFontVariantAlternatesValue::customCSSText const):
(WebCore::CSSFontVariantAlternatesValue::equals const):
* Source/WebCore/css/CSSFontVariantAlternatesValue.h: Added.
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::CSSPrimitiveValue::operator FontVariantAlternates const): Deleted.
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::visitDerived):
* Source/WebCore/css/CSSValue.h:
(WebCore::CSSValue::isFontVariantAlternatesValue const):
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::fontVariantAlternatesPropertyValue):
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::CSSParserFastPaths::isValidKeywordPropertyAndValue):
(WebCore::CSSParserFastPaths::isKeywordPropertyID):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeFontVariantAlternates):
(WebCore::CSSPropertyParser::parseSingleValue):
(WebCore::CSSPropertyParser::consumeFontVariantShorthand):
* Source/WebCore/platform/graphics/FontCascadeCache.h:
(WebCore::FontDescriptionKeyRareData::create):
(WebCore::FontDescriptionKeyRareData::variantAlternates const):
(WebCore::FontDescriptionKeyRareData::operator== const):
(WebCore::FontDescriptionKeyRareData::FontDescriptionKeyRareData):
(WebCore::add):
(WebCore::FontDescriptionKey::FontDescriptionKey):
(WebCore::FontDescriptionKey::makeFlagsKey):
* Source/WebCore/platform/graphics/FontCascadeDescription.cpp:
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
(WebCore::FontCascadeDescription::initialVariantAlternates):
* Source/WebCore/platform/graphics/FontDescription.cpp:
(WebCore::FontDescription::FontDescription):
* Source/WebCore/platform/graphics/FontDescription.h:
(WebCore::FontDescription::variantAlternates const):
(WebCore::FontDescription::setVariantAlternates):
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::computeFeatureSettingsFromVariants):
* Source/WebCore/platform/graphics/harfbuzz/ComplexTextControllerHarfBuzz.cpp:
(WebCore::setFeatureSettingsFromVariants):
* Source/WebCore/platform/text/TextFlags.cpp:
(WebCore::reduce):
(WebCore::operator&lt;&lt;):
(WebCore::add):
* Source/WebCore/platform/text/TextFlags.h:
(WebCore::FontVariantAlternates::isNormal const):
(WebCore::FontVariantAlternates::values const):
(WebCore::FontVariantAlternates::valuesRef):
(WebCore::FontVariantAlternates::setValues):
(WebCore::FontVariantAlternates::Normal):
(WebCore::FontVariantSettings::FontVariantSettings):
(WebCore::FontVariantSettings::isAllNormal const):
(WebCore::FontVariantSettings::uniqueValue const): Deleted.
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyInheritFontVariantAlternates):
(WebCore::Style::BuilderCustom::applyInitialFontVariantAlternates):
(WebCore::Style::BuilderCustom::applyValueFontVariantAlternates):

Canonical link: <a href="https://commits.webkit.org/255201@main">https://commits.webkit.org/255201@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3bf6da3b4be92af3c086fb95930c08a87143421e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91722 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/954 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/22312 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/101418 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/161491 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/95727 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/952 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29520 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/84039 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/97727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97380 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/566 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/78347 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/84039 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/82490 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/22312 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/84039 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35840 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/22312 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/33594 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/22312 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3604 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37440 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/78347 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39342 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/22312 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebCore/css/CSSFontVariantAlternatesValue.cpp:
```diff
@@ -0,0 +1,51 @@
+String CSSFontVariantAlternatesValue::customCSSText() const
+{
+    TextStream ts;
+    ts << this->m_value;
```
I don't think you need `this->`.

litherum commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -597,12 +597,9 @@
-            ],
             "codegen-properties": {
                 "name-for-methods": "VariantAlternates",
+                "custom": "All",
```
Are you sure this actually needs to be fully custom? Don't you just need `converter` instead?

mdubet replied:
I'm not sure, but:

- I dunno how to access the fontDescription from the `converter`.

- all the others FontVariant-(EastAsian/Numeric/...) are fully custom.

litherum replied:
Ok. I think it's worth a few minutes to try to determine what the technical reason that we can't use `converter` for all of these is (perhaps in a follow-up patch).

The purpose of the `converter` is so you don't have to access the `RenderStyle` (or the `FontDescription`) from the converter function. The converter function just turns the `CSSValue` into its backing type (which for us is `FontVariantAlternates`), and returns it. Then, generated code is responsible for assigning that value to the `RenderStyle` (or `FontDescription`).

litherum commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -820,9 +821,83 @@ static RefPtr<CSSPrimitiveValue> consumeFontVariantCaps(CSSParserTokenRange& ran
+static RefPtr<CSSValue> consumeFontVariantAlternates(CSSParserTokenRange& range)
 {
-    return consumeIdent<CSSValueNormal, CSSValueHistoricalForms>(range);
+    // return early if nothing to parse
```
Comments should answer the question "why" rather than the question "what".

litherum commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -820,9 +821,83 @@ static RefPtr<CSSPrimitiveValue> consumeFontVariantCaps(CSSParserTokenRange& ran
+    if (range.atEnd())
+        return nullptr;
+
+        // "normal" has to be single value.
```
I suggest deleting this comment. It doesn't provide any more information beyond what the code directly below it provides.

litherum commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -820,9 +821,83 @@ static RefPtr<CSSPrimitiveValue> consumeFontVariantCaps(CSSParserTokenRange& ran
+            if (token.id() == CSSValueHistoricalForms) {
+                consumeIdent<CSSValueHistoricalForms>(range);
+                
+                // Should not be parsed more than once.
```
Ditto.

litherum commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -820,9 +821,83 @@ static RefPtr<CSSPrimitiveValue> consumeFontVariantCaps(CSSParserTokenRange& ran
+                consumeIdent<CSSValueHistoricalForms>(range);
+                
+                // Should not be parsed more than once.
+                if (result.valuesRef().historicalForms)
+                    return false;
```
Is this tested?

mdubet replied:
Test added.

litherum commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -820,9 +821,83 @@ static RefPtr<CSSPrimitiveValue> consumeFontVariantCaps(CSSParserTokenRange& ran
+                if (result.valuesRef().historicalForms)
+                    return false;
+
+                // Otheriwse, set the value.
```
Ditto about the comment.

litherum commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -820,9 +821,83 @@ static RefPtr<CSSPrimitiveValue> consumeFontVariantCaps(CSSParserTokenRange& ran
+            if (!args.atEnd())
+                return false;
+        
+            // Value should contains a string.
```
Ditto about the comment.

litherum commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -820,9 +821,83 @@ static RefPtr<CSSPrimitiveValue> consumeFontVariantCaps(CSSParserTokenRange& ran
+            if (!ident)
+                return false;
+        
+            // Value should not be set multiple times.
```
Ditto.

litherum commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -820,9 +821,83 @@ static RefPtr<CSSPrimitiveValue> consumeFontVariantCaps(CSSParserTokenRange& ran
+            if (setter)
+                return false;
+        
+            // Everything is fine, set the value.
```
Ditto.

litherum commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -820,9 +821,83 @@ static RefPtr<CSSPrimitiveValue> consumeFontVariantCaps(CSSParserTokenRange& ran
+
+    auto parseSomethingWithoutError = [&range, &result]() {
+        bool hasParsedSomething = false;
+        auto parseAndSetArgument = [&range, &hasParsedSomething] (auto& setter) {
```
nit: I don't think `setter` is a great name when it's not a function.

mdubet replied:
Changed to `value`.

litherum commented on Source/WebCore/platform/graphics/FontCascadeCache.h:
```diff
@@ -43,9 +43,9 @@ namespace WebCore {
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    static Ref<FontDescriptionKeyRareData> create(FontFeatureSettings&& featureSettings, FontVariationSettings&& variationSettings, FontPalette&& fontPalette)
+    static Ref<FontDescriptionKeyRareData> create(FontFeatureSettings&& featureSettings, FontVariationSettings&& variationSettings, FontVariantAlternates&& variantAlternates, FontPalette&& fontPalette)
```
Why are the other `font-variant-*` properties not included in this? Why is `font-variant-alternates` special? This smells fishy.

mdubet replied:
The other ones are already included in the  `std::array<unsigned, 2> m_flags { { 0, 0 } }` (like alternates before this patch cf the change to the `makeFlagsKey` function).

litherum commented on Source/WebCore/platform/graphics/FontCascadeCache.h:
```diff
@@ -100,9 +109,10 @@ struct FontDescriptionKey {
     {
         auto featureSettings = description.featureSettings();
         auto variationSettings = description.variationSettings();
+        auto variantAlternates = description.variantAlternates();
```
Ditto.

litherum commented on Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
```diff
@@ -236,14 +236,13 @@ static FeaturesMap computeFeatureSettingsFromVariants(const FontVariantSettings&
+        if (values.historicalForms)
+            result.add(fontFeatureTag("hist"), 1);
+        
+        // TODO: handle other tags.
```
Bugzilla link?

mdubet replied:
Added https://bugs.webkit.org/show_bug.cgi?id=246121

litherum commented on Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
```diff
@@ -236,14 +236,13 @@ static FeaturesMap computeFeatureSettingsFromVariants(const FontVariantSettings&
-        break;
-    default:
-        ASSERT_NOT_REACHED();
+    // We don't do anything if "normal".
```
Ditto about the comment.

litherum commented on Source/WebCore/platform/graphics/harfbuzz/ComplexTextControllerHarfBuzz.cpp:
```diff
@@ -309,12 +309,13 @@ static void setFeatureSettingsFromVariants(const FontVariantSettings& variantSet
-    case FontVariantAlternates::HistoricalForms:
-        features.set(fontFeatureTag("hist"), 1);
-        break;
+    // We don't do anything if "normal".
```
Ditto about the comment.

litherum commented on Source/WebCore/platform/text/TextFlags.cpp:
```diff
@@ -58,15 +60,69 @@ WTF::TextStream& operator<<(TextStream& ts, Kerning kerning)
+template <typename T> using BinOp = std::function<T(T, T)>;
+
+template <typename T>
+T join(const std::vector<T>& vec, const BinOp<T>& binOp)
```
This is a classic `reduce()` function. I suggest moving this to `std::vector` right next to `map()`, naming it `reduce()`, and adding a 3rd argument `T initialAccumulator = T()`.

mdubet replied:
You mean moving it to `WTF::Vector` ?

litherum replied:
Yes

litherum commented on Source/WebCore/platform/text/TextFlags.cpp:
```diff
@@ -58,15 +60,69 @@ WTF::TextStream& operator<<(TextStream& ts, Kerning kerning)
+    if (vec.size() == 1)
+        return vec[0];
+    
+    auto acc = vec[0];
```
We generally spell out names using full words.

litherum commented on Source/WebCore/platform/text/TextFlags.cpp:
```diff
@@ -58,15 +60,69 @@ WTF::TextStream& operator<<(TextStream& ts, Kerning kerning)
+        if (values.annotation)
+            result.push_back("annotation(" + *values.annotation + ")");
+        
+        // Join all values with a space between them.
```
Again with the comment.

litherum commented on Source/WebCore/platform/text/TextFlags.h:
```diff
@@ -163,7 +166,64 @@ enum class FontVariantNumericFraction : uint8_t {
+    std::optional<String> annotation;
+    bool historicalForms = false;
+    friend void add(Hasher&, const FontVariantAlternatesValues&);
+};
```
Blank line between struct definitions, please

litherum commented on Source/WebCore/platform/text/TextFlags.h:
```diff
@@ -163,7 +166,64 @@ enum class FontVariantNumericFraction : uint8_t {
+    bool historicalForms = false;
+    friend void add(Hasher&, const FontVariantAlternatesValues&);
+};
+struct FontVariantAlternates {
```
If the members are private, it should be a class, and the members' names should start with `m_`.

litherum commented on Source/WebCore/platform/text/TextFlags.h:
```diff
@@ -163,7 +166,64 @@ enum class FontVariantNumericFraction : uint8_t {
+
+private:
+    std::variant<FontVariantAlternatesNormal, Values> val;
+    FontVariantAlternates() { };
```
`= default`

litherum commented on Source/WebCore/platform/text/TextFlags.h:
```diff
@@ -163,7 +166,64 @@ enum class FontVariantNumericFraction : uint8_t {
+};
+
+struct FontVariantAlternatesValues {
+    bool operator==(const FontVariantAlternatesValues&) const = default;
```
Can we add a blank line between functions and members?

litherum commented on Source/WebCore/platform/text/TextFlags.h:
```diff
@@ -163,7 +166,64 @@ enum class FontVariantNumericFraction : uint8_t {
+
+struct FontVariantAlternatesValues {
+    bool operator==(const FontVariantAlternatesValues&) const = default;
+    std::optional<String> stylistic;
+    std::optional<String> styleset;
+    std::optional<String> characterVariant;
+    std::optional<String> swash;
+    std::optional<String> ornaments;
+    std::optional<String> annotation;
```
Can these be enums rather than strings?

mdubet replied:
I'm not sure I understand this comment 🤔  Those value can be arbitrary strings.

litherum replied:
Ah, you're right, ok.

litherum commented on Source/WebCore/platform/text/TextFlags.cpp:
```diff
@@ -58,15 +60,69 @@ WTF::TextStream& operator<<(TextStream& ts, Kerning kerning)
+    if (alternates.isNormal())
+        ts << "normal";
+    else {
+        auto values = alternates.values();
+        std::vector<String> result;
+        if (values.stylistic)
+            result.push_back("stylistic(" + *values.stylistic + ")");
+        if (values.historicalForms)
+            result.push_back("historical-forms"_s);
+        if (values.styleset)
+            result.push_back("styleset(" + *values.styleset + ")");
+        if (values.characterVariant)
+            result.push_back("character-variant(" + *values.characterVariant + ")");
+        if (values.swash)
+            result.push_back("swash(" + *values.swash + ")");
+        if (values.ornaments)
+            result.push_back("ornaments(" + *values.ornaments + ")");
+        if (values.annotation)
+            result.push_back("annotation(" + *values.annotation + ")");
```
I kind of feel this is a layering violation, where the serialization rules here are defined in CSS, for CSS. This code is in platform/ and shouldn't know anything about CSS serialization rules. I think this function should be hoisted up to `CSSFontVariantAlternatesValue`.

mdubet replied:
It follows the CSS serialization rule because it's a convenient serialisation, but it's not mandatory indeed (because it's in platform).

I could move it to the css/ folder, but I would need roughly the exact same code here to implement the `<< stream` operator. WDYT ?

mdubet replied:
I've added a comment in the`CSSFontVariantAlternatesValue::customCSSText()`.

litherum replied:
ok

litherum commented on Source/WebCore/platform/graphics/FontCascadeCache.h:
```diff
@@ -142,10 +152,9 @@ struct FontDescriptionKey {
-        unsigned second = static_cast<unsigned>(description.variantEastAsianRuby()) << 27
-            | static_cast<unsigned>(description.variantEastAsianWidth()) << 25
-            | static_cast<unsigned>(description.variantEastAsianVariant()) << 22
-            | static_cast<unsigned>(description.variantAlternates()) << 21
```
I'd add a comment here saying that this was moved up to the Rare object. (Because variant-alternates can't be represented as just a bitfield)

litherum commented on Source/WebCore/platform/text/TextFlags.cpp:
```diff
@@ -58,15 +60,67 @@ WTF::TextStream& operator<<(TextStream& ts, Kerning kerning)
+template <typename T> using BinOp = std::function<T(T, T)>;
+
+template <typename T>
+T reduce(const std::vector<T>& vec, const BinOp<T>& binOp, std::optional<T> initial = { })
```
Oh, I think I led you astray with a previous comment. I meant for this to be a member function of `WTF::Vector` (when I said `std` earlier that was an accidental typo). It should live right next to `WTF::Vector::map()`. Such a general function doesn't belong in TextFlags.cpp.

This function could also be accompanied by a few tests in TestWTF with a few test cases.

Please feel free to do this part in a separate patch.

mdubet replied:
Yes, I was planning on moving this function to `WTF::Vector` that in a follow-up patch.

litherum reviewed (approved):
Looks good.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -826,9 +827,76 @@ static RefPtr<CSSPrimitiveValue> consumeFontVariantCaps(CSSParserTokenRange& ran
+            value = ident->stringValue();
+            return true;
+        };
+        for (;;) {
```
```suggestion
        while (true) {
```

would be better for consistency.

mdubet replied:
Fixed

mdubet committed:
Implement parsing of CSS font-variant-alternates
https://bugs.webkit.org/show_bug.cgi?id=149772
rdar://problem/22959741

Reviewed by Myles C. Maxfield.

This patch implements the complete parsing of the property.
Only the "normal" and "historical-forms" values are actually
passed down to the platform.

https://www.w3.org/TR/css-fonts-4/#font-variant-alternates-prop

* LayoutTests/fast/text/font-variant-shorthand-expected.txt:
* LayoutTests/fast/text/font-variant-shorthand.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-variant-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-001-expected.txt:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSFontVariantAlternatesValue.cpp: Added.
(WebCore::CSSFontVariantAlternatesValue::CSSFontVariantAlternatesValue):
(WebCore::CSSFontVariantAlternatesValue::customCSSText const):
(WebCore::CSSFontVariantAlternatesValue::equals const):
* Source/WebCore/css/CSSFontVariantAlternatesValue.h: Added.
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::CSSPrimitiveValue::operator FontVariantAlternates const): Deleted.
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::visitDerived):
* Source/WebCore/css/CSSValue.h:
(WebCore::CSSValue::isFontVariantAlternatesValue const):
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::fontVariantAlternatesPropertyValue):
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::CSSParserFastPaths::isValidKeywordPropertyAndValue):
(WebCore::CSSParserFastPaths::isKeywordPropertyID):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeFontVariantAlternates):
(WebCore::CSSPropertyParser::parseSingleValue):
(WebCore::CSSPropertyParser::consumeFontVariantShorthand):
* Source/WebCore/platform/graphics/FontCascadeCache.h:
(WebCore::FontDescriptionKeyRareData::create):
(WebCore::FontDescriptionKeyRareData::variantAlternates const):
(WebCore::FontDescriptionKeyRareData::operator== const):
(WebCore::FontDescriptionKeyRareData::FontDescriptionKeyRareData):
(WebCore::add):
(WebCore::FontDescriptionKey::FontDescriptionKey):
(WebCore::FontDescriptionKey::makeFlagsKey):
* Source/WebCore/platform/graphics/FontCascadeDescription.cpp:
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
(WebCore::FontCascadeDescription::initialVariantAlternates):
* Source/WebCore/platform/graphics/FontDescription.cpp:
(WebCore::FontDescription::FontDescription):
* Source/WebCore/platform/graphics/FontDescription.h:
(WebCore::FontDescription::variantAlternates const):
(WebCore::FontDescription::setVariantAlternates):
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::computeFeatureSettingsFromVariants):
* Source/WebCore/platform/graphics/harfbuzz/ComplexTextControllerHarfBuzz.cpp:
(WebCore::setFeatureSettingsFromVariants):
* Source/WebCore/platform/text/TextFlags.cpp:
(WebCore::reduce):
(WebCore::operator<<):
(WebCore::add):
* Source/WebCore/platform/text/TextFlags.h:
(WebCore::FontVariantAlternates::isNormal const):
(WebCore::FontVariantAlternates::values const):
(WebCore::FontVariantAlternates::valuesRef):
(WebCore::FontVariantAlternates::setValues):
(WebCore::FontVariantAlternates::Normal):
(WebCore::FontVariantSettings::FontVariantSettings):
(WebCore::FontVariantSettings::isAllNormal const):
(WebCore::FontVariantSettings::uniqueValue const): Deleted.
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyInheritFontVariantAlternates):
(WebCore::Style::BuilderCustom::applyInitialFontVariantAlternates):
(WebCore::Style::BuilderCustom::applyValueFontVariantAlternates):

Canonical link: https://commits.webkit.org/255201@main
